README.md
# SitemapGenerator

SitemapGenerator is the easiest way to generate Sitemaps in Ruby.  Rails integration provides access to the Rails route helpers within your sitemap config file and automatically makes the rake tasks available to you.  Or if you prefer to use another framework, you can!  You can use the rake tasks provided or run your sitemap configs as plain ruby scripts.

Sitemaps adhere to the [Sitemap 0.9 protocol][sitemap_protocol] specification.

## Features

* Framework agnostic
* Supports [News sitemaps][sitemap_news], [Video sitemaps][sitemap_video], [Image sitemaps][sitemap_images], [Mobile sitemaps][sitemap_mobile], [PageMap sitemaps][sitemap_pagemap] and [Alternate Links][alternate_links]
* Supports read-only filesystems like Heroku via uploading to a remote host like Amazon S3
* Compatible with all versions of Rails and Ruby
* Adheres to the [Sitemap 0.9 protocol][sitemap_protocol]
* Handles millions of links
* Customizable sitemap compression
* Notifies search engines (Google, Bing) of new sitemaps
* Ensures your old sitemaps stay in place if the new sitemap fails to generate
* Gives you complete control over your sitemap contents and naming scheme
* Intelligent sitemap indexing

### Show Me

This is a simple standalone example.  For Rails installation see the [Rails instructions](#rails) in the [Install](#installation) section.

Install:

```
gem install sitemap_generator
```

Create `sitemap.rb`:

```ruby
require 'rubygems'
require 'sitemap_generator'

SitemapGenerator::Sitemap.default_host = 'http://example.com'
SitemapGenerator::Sitemap.create do
  add '/home', :changefreq => 'daily', :priority => 0.9
  add '/contact_us', :changefreq => 'weekly'
end
SitemapGenerator::Sitemap.ping_search_engines # Not needed if you use the rake tasks
```

Run it:

```
ruby sitemap.rb
```

Output:

```
In /Users/karl/projects/sitemap_generator-test/public/
+ sitemap.xml.gz                                           3 links /  364 Bytes
Sitemap stats: 3 links / 1 sitemaps / 0m00s

Successful ping of Google
Successful ping of Bing
```

## Contents

* [Features](#features)
  + [Show Me](#show-me)
* [Contents](#contents)
* [Contribute](#contribute)
* [Foreword](#foreword)
* [Installation](#installation)
  + [Ruby](#ruby)
  + [Rails](#rails)
* [Getting Started](#getting-started)
  + [Preventing Output](#preventing-output)
  + [Rake Tasks](#rake-tasks)
  + [Pinging Search Engines](#pinging-search-engines)
  + [Crontab](#crontab)
  + [Robots.txt](#robotstxt)
  + [Ruby Modules](#ruby-modules)
  + [Deployments & Capistrano](#deployments--capistrano)
  + [Sitemaps with no Index File](#sitemaps-with-no-index-file)
  + [Upload Sitemaps to a Remote Host using Adapters](#upload-sitemaps-to-a-remote-host-using-adapters)
    - [Supported Adapters](#supported-adapters)
      * [`SitemapGenerator::FileAdapter`](#sitemapgeneratorfileadapter)
      * [`SitemapGenerator::FogAdapter`](#sitemapgeneratorfogadapter)
      * [`SitemapGenerator::S3Adapter`](#sitemapgenerators3adapter)
      * [`SitemapGenerator::AwsSdkAdapter`](#sitemapgeneratorawssdkadapter)
      * [`SitemapGenerator::WaveAdapter`](#sitemapgeneratorwaveadapter)
      * [`SitemapGenerator::GoogleStorageAdapter`](#sitemapgeneratorgooglestorageadapter)
    - [An Example of Using an Adapter](#an-example-of-using-an-adapter)
  + [Generating Multiple Sitemaps](#generating-multiple-sitemaps)
* [Sitemap Configuration](#sitemap-configuration)
  + [A Simple Example](#a-simple-example)
  + [Adding Links](#adding-links)
  + [Supported Options to `add`](#supported-options-to-add)
  + [Adding Links to the Sitemap Index](#adding-links-to-the-sitemap-index)
  + [Accessing the LinkSet instance](#accessing-the-linkset-instance)
  + [Speeding Things Up](#speeding-things-up)
* [Customizing your Sitemaps](#customizing-your-sitemaps)
  + [Sitemap Options](#sitemap-options)
* [Sitemap Groups](#sitemap-groups)
  + [A Groups Example](#a-groups-example)
  + [Using `group` without a block](#using-group-without-a-block)
* [Sitemap Extensions](#sitemap-extensions)
  + [News Sitemaps](#news-sitemaps)
    - [Example](#example)
    - [Supported options](#supported-options)
  + [Image Sitemaps](#image-sitemaps)
    - [Example](#example-1)
    - [Supported options](#supported-options-1)
  + [Video Sitemaps](#video-sitemaps)
    - [Example](#example-2)
    - [Supported options](#supported-options-2)
  + [PageMap Sitemaps](#pagemap-sitemaps)
    - [Supported options](#supported-options-3)
    - [Example:](#example)
  + [Alternate Links](#alternate-links)
    - [Example](#example-3)
    - [Supported options](#supported-options-4)
  + [Mobile Sitemaps](#-mobile-sitemaps)
    - [Example](#example-4)
    - [Supported options](#supported-options-5)
* [Compatibility](#compatibility)
* [Licence](#licence)

## Contribute

Does your website use SitemapGenerator to generate Sitemaps?  Where would you be without Sitemaps?  Probably still knocking rocks together.  Consider donating to the project to keep it up-to-date and open source.

<a href='http://www.pledgie.com/campaigns/15267'><img alt='Click here to lend your support to: SitemapGenerator and make a donation at www.pledgie.com !' src='http://pledgie.com/campaigns/15267.png?skin_name=chrome' border='0' /></a>


## Foreword

Adam Salter first created SitemapGenerator while we were working together in Sydney, Australia.  Unfortunately, he passed away in 2009.  Since then I have taken over development of SitemapGenerator.

Those who knew him know what an amazing guy he was, and what an excellent Rails programmer he was.  His passing is a great loss to the Rails community.

The canonical repository is: [http://github.com/kjvarga/sitemap_generator][canonical_repo]


## Installation

### Ruby

```
gem install 'sitemap_generator'
```

To use the rake tasks add the following to your `Rakefile`:

```ruby
require 'sitemap_generator/tasks'
```

The Rake tasks expect your sitemap to be at `config/sitemap.rb` but if you need to change that call like so: `rake sitemap:refresh CONFIG_FILE="path/to/sitemap.rb"`

### Rails

SitemapGenerator works with all versions of Rails and has been tested in Rails 2, 3 and 4.

Add the gem to your `Gemfile`:

```ruby
gem 'sitemap_generator'
```

Alternatively, if you are not using a `Gemfile` add the gem to your `config/application.rb` file config block:

```ruby
config.gem 'sitemap_generator'
```

Note: SitemapGenerator automatically loads its Rake tasks when used with Rails. You **do not need** to require the `sitemap_generator/tasks` file.

## Getting Started

### Preventing Output

To disable all non-essential output you can pass the `-s` option to Rake, for example `rake -s sitemap:refresh`, or set the environment variable `VERBOSE=false` when calling as a Ruby script.

To disable output in-code use the following:

```ruby
SitemapGenerator.verbose = false
```

### Rake Tasks

* `rake sitemap:install` will create a `config/sitemap.rb` file which is your sitemap configuration
  and contains everything needed to build your sitemap.  See
  [**Sitemap Configuration**](#sitemap-configuration) below for more information about how to
  define your sitemap.

* `rake sitemap:refresh` will create or rebuild your sitemap files as needed.  Sitemaps are
  generated into the `public/` folder and by default are named `sitemap.xml.gz`, `sitemap1.xml.gz`,
  `sitemap2.xml.gz`, etc.  As you can see, they are automatically GZip compressed for you.  In this case,
  `sitemap.xml.gz` is your sitemap "index" file.

  `rake sitemap:refresh` will output information about each sitemap that is written including its
  location, how many links it contains, and the size of the file.

### Pinging Search Engines

Using `rake sitemap:refresh` will notify Google and Bing to let them know that a new sitemap
is available.  To generate new sitemaps without notifying search engines, use `rake sitemap:refresh:no_ping`.

If you want to customize the hash of search engines you can access it at:

```ruby
SitemapGenerator::Sitemap.search_engines
```

Usually you would be adding a new search engine to ping.  In this case you can modify
the `search_engines` hash directly.  This ensures that when
`SitemapGenerator::Sitemap.ping_search_engines` is called, your new search engine will be included.

If you are calling `ping_search_engines` manually, then you can pass your new search engine
directly in the call, as in the following example:

```ruby
SitemapGenerator::Sitemap.ping_search_engines(newengine: 'http://newengine.com/ping?url=%s')
```

The key gives the name of the search engine, as a string or symbol, and the value is the full URL to ping, with a string interpolation that will be replaced by the CGI escaped sitemap index URL.  If you have any literal percent characters in your URL you need to escape them with `%%`.

If you are calling `SitemapGenerator::Sitemap.ping_search_engines` from outside of your sitemap config file, then you will need to set `SitemapGenerator::Sitemap.default_host` and any other options that you set in your sitemap config which affect the location of the sitemap index file.  For example:

```ruby
SitemapGenerator::Sitemap.default_host = 'http://example.com'
SitemapGenerator::Sitemap.ping_search_engines
```

Alternatively, you can pass in the full URL to your sitemap index, in which case we would have just the following:

```ruby
SitemapGenerator::Sitemap.ping_search_engines('http://example.com/sitemap.xml.gz')
```

### Crontab

To keep your sitemaps up-to-date, setup a cron job.  Make sure to pass the `-s` option to silence rake.  That way you will only get email if the sitemap build fails.

If you're using [Whenever](https://github.com/javan/whenever), your schedule would look something like this:

```ruby
# config/schedule.rb
every 1.day, :at => '5:00 am' do
  rake "-s sitemap:refresh"
end
```

### Robots.txt

You should add the URL of the sitemap index file to `public/robots.txt` to help search engines find your sitemaps.  The URL should be the complete URL to the sitemap index.  For example:

```
Sitemap: http://www.example.com/sitemap.xml.gz
```

### Ruby Modules

If you need to include a module (e.g. a rails helper), you must include it in the sitemap interpreter
class.  The part of your sitemap configuration that defines your sitemaps is run within an instance
of the `SitemapGenerator::Interpreter`:

```ruby
SitemapGenerator::Interpreter.send :include, RoutingHelper
```

### Deployments & Capistrano

To include the capistrano tasks just add the following to your Capfile:

```ruby
require 'capistrano/sitemap_generator'
```

Configurable options:

```ruby
set :sitemap_roles, :web # default
```

Available capistrano tasks:

```ruby
sitemap:create   #Create sitemaps without pinging search engines
sitemap:refresh  #Create sitemaps and ping search engines
sitemap:clean    #Clean up sitemaps in the sitemap path
```

  **Generate sitemaps into a directory which is shared by all deployments.**

  You can set your sitemaps path to your shared directory using the `sitemaps_path` option.  For example if we have a directory `public/shared/` that is shared by all deployments we can have our sitemaps generated into that directory by setting:

```ruby
SitemapGenerator::Sitemap.sitemaps_path = 'shared/'
```

### Sitemaps with no Index File

The sitemap index file is created for you on-demand, meaning that if you have a large site with more than one sitemap file, you will have a sitemap index file to reference those sitemap files.  If however you have a small site with only one sitemap file, you don't require an index and so no index will be created.  In both cases the index and sitemap file's name, respectively, is `sitemap.xml.gz`.

You may want to always create an index, even if you only have a small site.  Or you may never want to create an index.  For these cases, you can use the `create_index` option to control index creation.  You can read about this option in the Sitemap Options section below.

To always create an index:

```ruby
SitemapGenerator::Sitemap.create_index = true
```

To never create an index:

```ruby
SitemapGenerator::Sitemap.create_index = false
```
Your sitemaps will still be called `sitemap.xml.gz`, `sitemap1.xml.gz`, `sitemap2.xml.gz`, etc.

And the default "intelligent" behaviour:

```ruby
SitemapGenerator::Sitemap.create_index = :auto
```

### Upload Sitemaps to a Remote Host using Adapters

_This section needs better documentation.  Please consider contributing._

Sometimes it is desirable to host your sitemap files on a remote server, and point robots
and search engines to the remote files.  For example, if you are using a host like Heroku,
which doesn't allow writing to the local filesystem.  You still require *some* write access,
because the sitemap files need to be written out before uploading.  So generally a host will
give you write access to a temporary directory.  On Heroku this is `tmp/` within your application
directory.

#### Supported Adapters

##### `SitemapGenerator::FileAdapter`

  Standard adapter, writes out to a file.

##### `SitemapGenerator::FogAdapter`

  Uses `Fog::Storage` to upload to any service supported by Fog.

  You must `require 'fog'` in your sitemap config before using this adapter,
  or `require` another library that defines `Fog::Storage`.

##### `SitemapGenerator::S3Adapter`

  Uses `Fog::Storage` to upload to Amazon S3 storage.

  You must `require 'fog-aws'` in your sitemap config before using this adapter.

##### `SitemapGenerator::AwsSdkAdapter`

  Uses `Aws::S3::Resource` to upload to Amazon S3 storage.  Includes automatic detection of your AWS
  credentials using `Aws::Credentials`.

  You must `require 'aws-sdk'` in your sitemap config before using this adapter,
  or `require` another library that defines `Aws::S3::Resource` and `Aws::Credentials`.

  An example of using this adapter in your sitemap configuration:

  ```ruby
  SitemapGenerator::Sitemap.adapter = SitemapGenerator::AwsSdkAdapter.new('s3_bucket',
    aws_access_key_id: 'AKIAI3SW5CRAZBL4WSTA',
    aws_secret_access_key: 'asdfadsfdsafsadf',
    aws_region: 'us-east-1'
  )
  ```

##### `SitemapGenerator::WaveAdapter`

  Uses `CarrierWave::Uploader::Base` to upload to any service supported by CarrierWave, for example,
  Amazon S3, Rackspace Cloud Files, and MongoDB's GridF.

  You must `require 'carrierwave'` in your sitemap config before using this adapter,
  or `require` another library that defines `CarrierWave::Uploader::Base`.

  Some documentation exists [on the wiki page][remote_hosts].

##### `SitemapGenerator::GoogleStorageAdapter`

  Uses [`Google::Cloud::Storage`][google_cloud_storage_gem] to upload to Google Cloud storage.

  You must `require 'google/cloud/storage'` in your sitemap config before using this adapter.

  An example of using this adapter in your sitemap configuration with options:

  ```ruby
  SitemapGenerator::Sitemap.adapter = SitemapGenerator::GoogleStorageAdapter.new(
    credentials: 'path/to/keyfile.json',
    project_id: 'google_account_project_id',
    bucket: 'name_of_bucket'
  )
  ```
  Also, inline with Google Authentication options, it can also pick credentials from environment variables. All [supported environment variables][google_cloud_storage_authentication] can be used, for example: `GOOGLE_CLOUD_PROJECT` and `GOOGLE_CLOUD_CREDENTIALS`.  An example of using this adapter with the environment variables is:

  ```ruby
  SitemapGenerator::Sitemap.adapter = SitemapGenerator::GoogleStorageAdapter.new(
    bucket: 'name_of_bucket'
  )
  ```

  All options other than the `:bucket` option are passed to the `Google::Cloud::Storage.new` initializer giving you maximum configurability.  See the [Google Cloud Storage initializer][google_cloud_storage_initializer] for supported options.

#### An Example of Using an Adapter

1. Please see [this wiki page][remote_hosts] for more information about setting up SitemapGenerator to upload to a
   remote host.

2. This example uses the CarrierWave adapter.  It shows some common settings that are used when the hostname hosting
   the sitemaps differs from the hostname of the sitemap links.

     ```ruby
     # Your website's host name
     SitemapGenerator::Sitemap.default_host = "http://www.example.com"

     # The remote host where your sitemaps will be hosted
     SitemapGenerator::Sitemap.sitemaps_host = "http://s3.amazonaws.com/sitemap-generator/"

     # The directory to write sitemaps to locally
     SitemapGenerator::Sitemap.public_path = 'tmp/'

     # Set this to a directory/path if you don't want to upload to the root of your `sitemaps_host`
     SitemapGenerator::Sitemap.sitemaps_path = 'sitemaps/'

     # The adapter to perform the upload of sitemap files.
     SitemapGenerator::Sitemap.adapter = SitemapGenerator::WaveAdapter.new
     ```

3. Update your `robots.txt` file to point robots to the remote sitemap index file, e.g:

    ```
    Sitemap: http://s3.amazonaws.com/sitemap-generator/sitemaps/sitemap.xml.gz
    ```

    You generate your sitemaps as usual using `rake sitemap:refresh`.

    Note that SitemapGenerator will automatically turn off `include_index` in this case because
    the `sitemaps_host` does not match the `default_host`.  The link to the sitemap index file
    that would otherwise be included would point to a different host than the rest of the links
    in the sitemap, something that the sitemap rules forbid.  (Since version 3.2 this is no
    longer an issue because [`include_index` is off by default][include_index_change].)

4. Verify to Google that you own the S3 url

    In order for Google to use your sitemap, you need to prove you own the S3 bucket through [google webmaster tools](https://www.google.com/webmasters/tools/home?hl=en).  In the example above, you would add the site `http://s3.amazonaws.com/sitemap-generator/sitemaps`.  Once you have verified you own the directory, then add your
    sitemap index to the list of sitemaps for the site.

### Generating Multiple Sitemaps

Each call to `create` creates a new sitemap index and associated sitemaps.  You can call `create` as many times as you want within your sitemap configuration.

You must remember to use a different filename or location for each set of sitemaps, otherwise they will
overwrite each other.  You can use the `filename`, `namer` and `sitemaps_path` options for this.

In the following example we generate three sitemaps each in its own subdirectory:

```ruby
%w(google bing apple).each do |subdomain|
  SitemapGenerator::Sitemap.default_host = "https://#{subdomain}.mysite.com"
  SitemapGenerator::Sitemap.sitemaps_path = "sitemaps/#{subdomain}"
  SitemapGenerator::Sitemap.create do
    add '/home'
  end
end
```

Outputs:

```
+ sitemaps/google/sitemap1.xml.gz             2 links /  822 Bytes /  328 Bytes gzipped
+ sitemaps/google/sitemap.xml.gz           1 sitemaps /  389 Bytes /  217 Bytes gzipped
Sitemap stats: 2 links / 1 sitemaps / 0m00s
+ sitemaps/bing/sitemap1.xml.gz               2 links /  820 Bytes /  330 Bytes gzipped
+ sitemaps/bing/sitemap.xml.gz             1 sitemaps /  388 Bytes /  217 Bytes gzipped
Sitemap stats: 2 links / 1 sitemaps / 0m00s
+ sitemaps/apple/sitemap1.xml.gz              2 links /  820 Bytes /  330 Bytes gzipped
+ sitemaps/apple/sitemap.xml.gz            1 sitemaps /  388 Bytes /  214 Bytes gzipped
Sitemap stats: 2 links / 1 sitemaps / 0m00s
```

If you don't want to have to generate all the sitemaps at once, or you want to refresh some more often than others, you can split them up into their own configuration files.  Using the above example we would have:

```ruby
# config/google_sitemap.rb
SitemapGenerator::Sitemap.default_host = "https://google.mysite.com"
SitemapGenerator::Sitemap.sitemaps_path = "sitemaps/google"
SitemapGenerator::Sitemap.create do
  add '/home'
end

# config/apple_sitemap.rb
SitemapGenerator::Sitemap.default_host = "https://apple.mysite.com"
SitemapGenerator::Sitemap.sitemaps_path = "sitemaps/apple"
SitemapGenerator::Sitemap.create do
  add '/home'
end

# config/bing_sitemap.rb
SitemapGenerator::Sitemap.default_host = "https://bing.mysite.com"
SitemapGenerator::Sitemap.sitemaps_path = "sitemaps/bing"
SitemapGenerator::Sitemap.create do
  add '/home'
end
```


To generate each one specify the configuration file to run by passing the `CONFIG_FILE` option to `rake sitemap:refresh`, e.g.:

```
rake sitemap:refresh CONFIG_FILE="config/google_sitemap.rb"
rake sitemap:refresh CONFIG_FILE="config/apple_sitemap.rb"
rake sitemap:refresh CONFIG_FILE="config/bing_sitemap.rb"
```

## Sitemap Configuration

A sitemap configuration file contains all the information needed to generate your sitemaps.  By default SitemapGenerator looks for a configuration file in  `config/sitemap.rb` - relative to your application root or the current working directory.  (Run `rake sitemap:install` to have this file generated for you if you have not done so already.)

If you want to use a non-standard configuration file, or have multiple configuration files, you can specify which one to run by passing the `CONFIG_FILE` option like so:

```
rake sitemap:refresh CONFIG_FILE="config/geo_sitemap.rb"
```

### A Simple Example

So what does a sitemap configuration look like?  Let's take a look at a simple example:

```ruby
SitemapGenerator::Sitemap.default_host = "http://www.example.com"
SitemapGenerator::Sitemap.create do
  add '/welcome'
end
```

A few things to note:

* `SitemapGenerator::Sitemap` is a lazy-initialized sitemap object provided for your convenience.
* Every sitemap must set `default_host`.  This is the hostname that is used when building links to add to the sitemap (and all links in a sitemap must belong to the same host).
* The `create` method takes a block with calls to `add` to add links to the sitemap.
* The sitemaps are written to the `public/` directory in the directory from which the script is run.  You can specify a custom location using the `public_path` or `sitemaps_path` option.

Now let's see what is output when we run this configuration with `rake sitemap:refresh:no_ping`:

```
In /Users/karl/projects/sitemap_generator-test/public/
+ sitemap.xml.gz                                           2 links /  347 Bytes
Sitemap stats: 2 links / 1 sitemaps / 0m00s
```

Weird!  The sitemap has two links, even though we only added one!  This is because SitemapGenerator adds the root URL `/` for you by default.  (Note that prior to version 3.2 the  URL of the sitemap index file was also added to the sitemap by default but [this behaviour has been changed][include_index_change] because of Google complaining about nested indexing.  This also doesn't make sense anymore because indexes are not always needed.)  You can change the default behaviour by setting the `include_root` or `include_index` option.

Now let's take a look at the file that was created.  After uncompressing and XML-tidying the contents we have:


* `public/sitemap.xml.gz`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
  <url>
    <loc>http://www.example.com/</loc>
    <lastmod>2011-05-21T00:03:38+00:00</lastmod>
    <changefreq>always</changefreq>
    <priority>1.0</priority>
  </url>
  <url>
    <loc>http://www.example.com/welcome</loc>
    <lastmod>2011-05-21T00:03:38+00:00</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.5</priority>
  </url>
</urlset>
```

The sitemaps conform to the [Sitemap 0.9 protocol][sitemap_protocol].  Notice the value for `priority` and `changefreq` on the root link, the one that was added for us?  The values tell us that this link is the highest priority and should be checked regularly because it are constantly changing.  You can specify your own values for these options in your call to `add`.

In this example no sitemap index was created because we have so few links, so none was needed.  If we run the same example above and set `create_index = true` we can take a look at what an index file looks like:

```ruby
SitemapGenerator::Sitemap.default_host = "http://www.example.com"
SitemapGenerator::Sitemap.create_index = true
SitemapGenerator::Sitemap.create do
  add '/welcome'
end
```

And the output:

```
In /Users/karl/projects/sitemap_generator-test/public/
+ sitemap1.xml.gz                                          2 links /  347 Bytes
+ sitemap.xml.gz                                        1 sitemaps /  228 Bytes
Sitemap stats: 2 links / 1 sitemaps / 0m00s
```

Now if we look at the uncompressed and formatted contents of `sitemap.xml.gz` we can see that it is a sitemap index and `sitemap1.xml.gz` is a sitemap:

* `public/sitemap.xml.gz`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/siteindex.xsd">
  <sitemap>
    <loc>http://www.example.com/sitemap1.xml.gz</loc>
    <lastmod>2013-05-01T18:10:26-07:00</lastmod>
  </sitemap>
</sitemapindex>
```

### Adding Links

You call `add` in the block passed to `create` to add a **path** to your sitemap.  `add` takes a string path and optional hash of options, generates the URL and adds it to the sitemap.  You only need to pass a **path** because the URL will be built for us using the `default_host` we specified.  However, if we want to use a different host for a particular link, we can pass the `:host` option to `add`.

Let's see another example:

```ruby
SitemapGenerator::Sitemap.default_host = "http://www.example.com"
SitemapGenerator::Sitemap.create do
  add '/contact_us'
  Content.find_each do |content|
    add content_path(content), :lastmod => content.updated_at
  end
end
```

In this example first we add the `/contact_us` page to the sitemap and then we iterate through the Content model's records adding each one to the sitemap using the `content_path` helper method to generate the path for each record.

The **Rails URL/path helper methods are automatically made available** to us in the `create` block.  This keeps the logic for building our paths out of the sitemap config and in the Rails application where it should be.  You use those methods just like you would in your application's view files.

In the example about we pass a `lastmod` (last modified) option with the value of the record's `updated_at` attribute so that search engines know to only re-index the page when the record changes.

Looking at the output from running this sitemap, we see that we have a few more links than before:

```

CHANGES.md
### 6.1.1

* Resolve deprecation warning on using Kernel#open in Ruby 2.7 (use URI.open instead) [#342](https://github.com/kjvarga/sitemap_generator/pull/342)
* Support S3 Endpoints for S3 Compliant Providers like DigitalOcean Spaces [#325](https://github.com/kjvarga/sitemap_generator/pull/325)

### 6.1.0

* Support uploading files to Google Cloud Storage [#326](https://github.com/kjvarga/sitemap_generator/pull/326) and [#340](https://github.com/kjvarga/sitemap_generator/pull/340)

### 6.0.2

* Resolve `BigDecimal.new is deprecated` warnings in Ruby 2.5 [#305](https://github.com/kjvarga/sitemap_generator/pull/305).
* Resolve `instance variable not initialized`, `File.exists? is deprecated` and `'*' interpreted as argument prefix` warnings [#304](https://github.com/kjvarga/sitemap_generator/pull/304).

### 6.0.1

* Use `yaml_tag` instead of `yaml_as`, which was deprecated in Ruby 2.4, and removed in 2.5 [#298](https://github.com/kjvarga/sitemap_generator/pull/298).

### 6.0.0

*Backwards incompatible changes*

* Adapters (AWS SDK, S3, Fog & Wave) no longer load their dependencies.  It is up to the user
  to `require` the appropriate libraries for the adapter to work.
* AwsSdkAdapter: Fixed [#279](https://github.com/kjvarga/sitemap_generator/issues/279) where sitemaps were incorrectly nested under a `sitemaps/` directory in S3
* Stop supporting Ruby < 2.0, test with Ruby 2.4.

*Other changes*

* If Rails is defined but the application is not loaded, don't include the URL helpers.

### 5.3.1

* Ensure files have 644 permissions when building to try to address issue [#264](https://github.com/kjvarga/sitemap_generator/issues/264)
* Use HTTPS in the Gemfile (PR #[#263](https://github.com/kjvarga/sitemap_generator/pull/263))

### 5.3.0

* Add `max_sitemap_links` option support for limiting how many links each sitemap can hold.  Issue [#188](https://github.com/kjvarga/sitemap_generator/issues/188) PR [#262](https://github.com/kjvarga/sitemap_generator/pull/262)
* Upgrade development dependencies
* Modernize Gemfile & gemspec
* Bring specs up to RSpec 3.5
* Remove Geo sitemap support.  Google no longer supports them. Issue [#246](https://github.com/kjvarga/sitemap_generator/issues/246)
* Use `sitemap` namespace for Capistrano tasks (rather than `deploy`). PR [#241](https://github.com/kjvarga/sitemap_generator/pull/241)
* Use presence of `Rails::VERSION` to detect when running under Rails, rather than just `Rails` constant.  PR [#221](https://github.com/kjvarga/sitemap_generator/pull/221)
* Remove gem post-install message warning about incompatible changes in version 4

### 5.2.0

* New `SitemapGenerator::AwsSdkAdapter` adapter using the bare aws-sdk gem.
* Fix Bing ping url.
* Support string option keys passed to `add`.
* In Railtie, Load the rake task instead of requiring them.

### 5.1.0

* Require only `fog-aws` instead of `fog` for the `S3Adapter` and support using IAM profile instead of setting access key & secret directly.
* Implement `respond_to?` on the `SitemapGenerator::Sitemap` pseudo class.
* Make `:lang` optional on alternate links so they can be used for [AppIndexing](https://developers.google.com/app-indexing/reference/deeplinks).
* Documented Mobile Sitemaps `:mobile` option.

### 5.0.5

* Use MIT licence.
* Fix deploys with Capistrano 3 ([#163](https://github.com/kjvarga/sitemap_generator/issues/163)).
* Allow any Fog storage options for S3 adapter ([#167](https://github.com/kjvarga/sitemap_generator/pull/167)).

### 5.0.4

* Don't include the `media` attribute on alternate links unless it's given

### 5.0.3

* Add support for Video sitemaps options `:live` and ':requires_subscription'

### 5.0.2

* Set maximum filesize to 10,000,000 bytes rather than 10,485,760 bytes.

### 5.0.1

* Include new `SitemapGenerator::FogAdapter` ([#138](https://github.com/kjvarga/sitemap_generator/pull/138)).
* Fix usage of attr_* methods in `LinkSet`
* Don't override custom getters/setters ([#144](https://github.com/kjvarga/sitemap_generator/pull/144)).
* Fix breaking spec in Ruby 2 ([#142](https://github.com/kjvarga/sitemap_generator/pull/142)).
* Include Capistrano 3.x tasks ([#141](https://github.com/kjvarga/sitemap_generator/pull/141)).

### 5.0.0

* Support new `:compress` option for customizing which files get compressed.
* Remove old deprecated methods:
    * Removed options to `LinkSet::add()`: `:sitemaps_namer` and `:sitemap_index_namer` (use `:namer` option)
    * Removed `LinkSet::sitemaps_namer=`, `LinkSet::sitemaps_namer` (use `LinkSet::namer=` and `LinkSet::namer`)
    * Removed `LinkSet::sitemaps_index_namer=`, `LinkSet::sitemaps_index_namer` (use `LinkSet::namer=` and `LinkSet::namer`)
    * Removed the `SitemapGenerator::SitemapNamer` class (use `SitemapGenerator::SimpleNamer`)
    * Removed `LinkSet::add_links()` (use `LinkSet::create()`)
* Support `fog_path_style` option in the `SitemapGenerator::S3Adapter` so buckets with dots in the name work over HTTPS without SSL certificate problems.

### 4.3.1

* Support integer timestamps.
* Update README for new features added in last release.

### 4.3.0

* Support `media` attibute on alternate links ([#125](https://github.com/kjvarga/sitemap_generator/issues/125)).
* Changed `SitemapGenerator::S3Adapter` to write files in a single operation, avoiding potential permissions errors when listing a directory prior to writing ([#130](https://github.com/kjvarga/sitemap_generator/issues/130)).
* Remove Sitemap Writer from ping task ([#129](https://github.com/kjvarga/sitemap_generator/issues/129)).
* Support `url:expires` element ([#126](https://github.com/kjvarga/sitemap_generator/issues/126)).

### 4.2.0

* Update Google ping URL.
* Quote the ping URL in the output.
* Support Video `video:price` element ([#117](https://github.com/kjvarga/sitemap_generator/issues/117)).
* Support symbols as well as strings for most arguments to `add()` ([#113](https://github.com/kjvarga/sitemap_generator/issues/113)).

lib/sitemap_generator.rb
require 'sitemap_generator/simple_namer'
require 'sitemap_generator/builder'
require 'sitemap_generator/link_set'
require 'sitemap_generator/templates'
require 'sitemap_generator/utilities'
require 'sitemap_generator/application'
require 'sitemap_generator/sitemap_location'

module SitemapGenerator
  autoload(:Interpreter,   'sitemap_generator/interpreter')
  autoload(:FileAdapter,   'sitemap_generator/adapters/file_adapter')
  autoload(:S3Adapter,     'sitemap_generator/adapters/s3_adapter')
  autoload(:AwsSdkAdapter, 'sitemap_generator/adapters/aws_sdk_adapter')
  autoload(:WaveAdapter,   'sitemap_generator/adapters/wave_adapter')
  autoload(:FogAdapter,    'sitemap_generator/adapters/fog_adapter')
  autoload(:BigDecimal,    'sitemap_generator/core_ext/big_decimal')
  autoload(:Numeric,       'sitemap_generator/core_ext/numeric')

  SitemapError          = Class.new(StandardError)
  SitemapFullError      = Class.new(SitemapError)
  SitemapFinalizedError = Class.new(SitemapError)

  Utilities.with_warnings(nil) do
    VERSION = File.read(File.dirname(__FILE__) + "/../VERSION").strip
    MAX_SITEMAP_FILES    = 50_000        # max sitemap links per index file
    MAX_SITEMAP_LINKS    = 50_000        # max links per sitemap
    MAX_SITEMAP_IMAGES   = 1_000         # max images per url
    MAX_SITEMAP_NEWS     = 1_000         # max news sitemap per index_file
    MAX_SITEMAP_FILESIZE = 50_000_000    # bytes
    SCHEMAS = {
      'image'   => 'http://www.google.com/schemas/sitemap-image/1.1',
      'mobile'  => 'http://www.google.com/schemas/sitemap-mobile/1.0',
      'news'    => 'http://www.google.com/schemas/sitemap-news/0.9',
      'pagemap' => 'http://www.google.com/schemas/sitemap-pagemap/1.0',
      'video'   => 'http://www.google.com/schemas/sitemap-video/1.1'
    }

    # Lazy-initialize the LinkSet instance
    Sitemap = (Class.new do
      def method_missing(*args, &block)
        (@link_set ||= reset!).send(*args, &block)
      end

      def respond_to?(name, include_private = false)
        (@link_set ||= reset!).respond_to?(name, include_private)
      end

      # Use a new LinkSet instance
      def reset!
        @link_set = LinkSet.new
      end
    end).new
  end

  class << self
    attr_accessor :root, :app, :templates
    attr_writer :yield_sitemap, :verbose
  end

  # Global default for the verbose setting.
  def self.verbose
    if @verbose.nil?
      @verbose = if SitemapGenerator::Utilities.truthy?(ENV['VERBOSE'])
        true
      elsif SitemapGenerator::Utilities.falsy?(ENV['VERBOSE'])
        false
      else
        nil
      end
    else
      @verbose
    end
  end

  # Returns true if we should yield the sitemap instance to the block, false otherwise.
  def self.yield_sitemap?
    !!@yield_sitemap
  end

  self.root      = File.expand_path(File.join(File.dirname(__FILE__), '../'))  # Root of the install dir, not the Rails app
  self.templates = SitemapGenerator::Templates.new(self.root)
  self.app       = SitemapGenerator::Application.new
end

require 'sitemap_generator/railtie' if SitemapGenerator.app.is_at_least_rails3?
